win32/tests/file_win32_ut/file_win32_ut.c

```diff
@@ -0,0 +1,952 @@
+    STRICT_EXPECTED_CALL(mock_CloseHandle(IGNORED_ARG));
+    STRICT_EXPECTED_CALL(mock_CloseThreadpoolIo(IGNORED_ARG))
+        .ValidateArgumentValue_pio(&captured_ptp_io);
+    STRICT_EXPECTED_CALL(free(file_handle));
+
+    ///act
+    file_destroy(file_handle);
+
+    ///assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+}
+
+/*Tests_SRS_FILE_43_009: [ If handle is NULL then file_write_async shall fail and return FILE_WRITE_ASYNC_INVALID_ARGS. ]*/
+/*Tests_SRS_FILE_WIN32_43_043: [ If handle is NULL then file_write_async shall fail and return FILE_WRITE_ASYNC_INVALID_ARGS. ]*/
+TEST_FUNCTION(file_write_async_fails_with_null_handle)
+{
+    ///arrange
+    unsigned char source[10];
+    FILE_CB user_callback = (FILE_CB)42;
+
+    ///act
+    FILE_WRITE_ASYNC_RESULT result = file_write_async(NULL, source, 1, 1, user_callback, NULL);
+
+    ///assert
+    ASSERT_ARE_EQUAL(FILE_WRITE_ASYNC_RESULT, FILE_WRITE_ASYNC_INVALID_ARGS, result);
+}
+
+/*Tests_SRS_FILE_43_010: [ If source is NULL then file_write_async shall fail and return FILE_WRITE_ASYNC_INVALID_ARGS. ]*/
+/*Tests_SRS_FILE_WIN32_43_044: [ If source is NULL then file_write_async shall fail and return FILE_WRITE_ASYNC_INVALID_ARGS. ]*/
+TEST_FUNCTION(file_write_async_fails_with_null_source)
+{
+    ///arrange
+    FILE_HANDLE file_handle = (FILE_HANDLE)42;
+    FILE_CB user_callback = (FILE_CB)43;
+
+    ///act
+    FILE_WRITE_ASYNC_RESULT result = file_write_async(file_handle, NULL, 1, 1, user_callback, NULL);
+
+    ///assert
+    ASSERT_ARE_EQUAL(FILE_WRITE_ASYNC_RESULT, FILE_WRITE_ASYNC_INVALID_ARGS, result);
+}
+
+/*Tests_SRS_FILE_43_012: [ If user_callback is NULL then file_write_async shall fail and return FILE_WRITE_ASYNC_INVALID_ARGS. ]*/
+/*Tests_SRS_FILE_WIN32_43_045: [ If user_callback is NULL then file_write_async shall fail and return FILE_WRITE_ASYNC_INVALID_ARGS. ]*/
+TEST_FUNCTION(file_write_async_fails_with_null_user_callback)
+{
+    ///arrange
+    FILE_HANDLE file_handle = (FILE_HANDLE)42;
+    unsigned char source[10];
+
+    ///act
+    FILE_WRITE_ASYNC_RESULT result = file_write_async(file_handle, source, 1, 1, NULL, NULL);
+
+    ///assert
+    ASSERT_ARE_EQUAL(FILE_WRITE_ASYNC_RESULT, FILE_WRITE_ASYNC_INVALID_ARGS, result);
+}
+
+/*Tests_SRS_FILE_43_040: [ If position + size is greater than INT64_MAX, then file_write_async shall fail and return FILE_WRITE_ASYNC_INVALID_ARGS. ]*/
+/*Tests_SRS_FILE_WIN32_43_060: [ If position + size is greater than INT64_MAX, then file_write_async shall fail and return FILE_WRITE_ASYNC_INVALID_ARGS. ]*/
+TEST_FUNCTION(file_write_async_fails_if_write_overflows_max_size)
+{
+    ///arrange
+    FILE_HANDLE file_handle = (FILE_HANDLE)42;
+    FILE_CB user_callback = (FILE_CB)43;
+    unsigned char source[10];
+
+    ///act
+    FILE_WRITE_ASYNC_RESULT result = file_write_async(file_handle, source, 1, INT64_MAX, user_callback, NULL);

```

Comment:
> 1, INT64_MAX [](start = 74, length = 13)

Ok, the more I read this, I keep forgetting it's "size" then "position". I think that all the arguments should maybe be reversed (position, size, buffer). Because offset/position then size is what I assume, and the buffer should be next to the size.

I don't feel super strongly about it, but it does look a little backwards to me (but it's kind of a pain to reverse everything now so leaving that up to you to decide).

Reply:
make a task

---
In reply to: [456720329](https://github.com/Azure/azure-c-pal/pull/10#discussion_r456720329) [](ancestors = 456720329)